compile.py

```diff
@@ -0,0 +1,608 @@
+
+INDENT = 2
+
+class CantCompile(Exception):
+    def __init__(self, tree, hint=None):
+        super().__init__(f"Could not compile `{ast.unparse(tree)}`")
+        self.tree = tree
+        self.hint = hint
+        self.msg = None
+
+ISSUES = []
+
+def catch_issues(f):
+    def wrapped(tree, *args, **kwargs):
+        try:
+            return f(tree, *args, **kwargs)
+        except AssertionError as e2:
+            try:
+                return find_hint(tree, "js")
+            except CantCompile as e:
+                e.msg = str(e2)
+                ISSUES.append(e)
+                return "/* " + ast.unparse(tree) + " */"
+        except CantCompile as e:
+            if not e.hint:
+                try:
+                    return find_hint(tree, "js")
+                except CantCompile as e2:
+                    e = e2
+            ISSUES.append(e)
+            return "/* " + ast.unparse(e.tree) + " */"
+    return wrapped
+
+HINTS = []
+
+def read_hints(f):
+    global HINTS
+    hints = json.load(f)
+    for h in hints:
+        assert "line" in h
+        assert isinstance(h["line"], int)
+        assert "code" in h
+        s = ast.parse(h["code"])
+        assert isinstance(s, ast.Module)
+        assert len(s.body) == 1
+        assert isinstance(s.body[0], ast.Expr)
+        h["ast"] = s.body[0].value
+        h["used"] = False
+    HINTS = hints
+    
+def find_hint(t, key):
+    for h in HINTS:
+        if h["line"] != t.lineno: continue
+        if ast.dump(h["ast"]) != ast.dump(t): continue
+        if key not in h: continue
+        break
+    else:
+        raise CantCompile(t, hint={"line": t.lineno, "code": ast.unparse(t), key: "???"})
+    assert not h["used"]
+    h["used"] = True
+    return h[key]
+
+def check_args(args, ctx):
+    assert not args.posonlyargs, ast.dump(args)
+    assert not args.vararg, ast.dump(args)
+    assert not args.kwonlyargs, ast.dump(args)
+    assert not args.kw_defaults, ast.dump(args)
+    assert not args.kwarg, ast.dump(args)
+    out = []
+    defaults = ([None] * len(args.args) + args.defaults)[-len(args.args):]
+    for i, (arg, default) in enumerate(zip(args.args, defaults)):
+        assert not arg.annotation
+        assert not arg.type_comment
+        if ctx.type == "class" and i == 0:
+            assert arg.arg == "self"
+        else:
+            if default:
+                out.append(arg.arg + " = " + compile_expr(default, ctx))
+            else:
+                out.append(arg.arg)
+    return out
+
+RENAME_METHODS = {
+    "lower": "toLowerCase",
+    "upper": "toUpperCase",
+    "strip": "trim",
+    "append": "push",
+    "pop": "pop",
+    "startswith": "startsWith",
+}
+
+RENAME_FNS = {
+    "int": "Math.parseInt",
+    "float": "Math.parseFloat",
+    "print": "console.log",
+}
+
+IMPORTS = []
+
+LIBRARY_METHODS = [
+    # socket
+    "connect",
+    "wrap_socket",
+    "send",
+    "makefile",
+    "readline",
+    "read",
+    #"close", # Shadows our code
```

Comment:
Yeah, this is annoying. The compiler needs to know which methods are "ours" and which are from libraries, because libraries need to be called without `async`. (Do they? Maybe not, actually?) For convenience, it assumes all method names are one or the other. Except our code defines a `close` method, as does the `socket` library. We should rename ours to `close_tag` because it's a bad name anyway, which will make this issue go away.